src/common/lib/util/defaults.ts

```diff
@@ -262,16 +263,15 @@ export function normaliseOptions(
 
   if (!('queueMessages' in options)) options.queueMessages = true;
 
-  /* infer hosts and fallbacks based on the configured environment */
-  const environment = (options.environment && String(options.environment).toLowerCase()) || Defaults.ENVIRONMENT;
-  const production = !environment || environment === 'production';
+  /* infer hosts and fallbacks based on the specified endpoint */
+  const endpoint = options.endpoint || options.environment || Defaults.ENDPOINT;
 
   if (!options.fallbackHosts && !options.restHost && !options.realtimeHost && !options.port && !options.tlsPort) {
-    options.fallbackHosts = production ? Defaults.FALLBACK_HOSTS : environmentFallbackHosts(environment);
+    options.fallbackHosts = getEndpointFallbackHosts(endpoint);
   }
 
-  const restHost = options.restHost || (production ? Defaults.REST_HOST : environment + '-' + Defaults.REST_HOST);
-  const realtimeHost = getRealtimeHost(options, production, environment, loggerToUse);
+  const restHost = options.restHost || getEndpointHostname(endpoint);
```

Comment:
I don't think that the spec has an internal concept of separate "rest host" and "realtime host" any more; isn't there just the "primary domain"? And then, REC1d specifies how the `restHost` and `realtimeHost` options, if specified, determine the primary domain.

Reply:
I've pushed 65a30d5ebb05d571737c9f9723d61d457078508f, is that what you meant? I don't think I want to go and change `restHost` and `realtimeHost` anywhere else internally (or have the capacity to do that).

Reply:
I think that if you don't feel comfortable making any further-reaching changes, then the change you've made is a good one.